File: packages/splitTokens-js/src/splitTokens.js
import normalizeInput from './normalizeInput.js';
import parseMultiLine from './parseMultiLine.js';
import parseSingleLine from './parseSingleLine.js';

/**
 * 汎用入力パーサ
 * - 半角/全角スペース, タブ, カンマ, 改行に対応（空白・改行は normalizeInput で正規化）
 * - デフォルトは「単一値でも配列」で返す
 * - オプション primitive: true のときのみ、単一行かつ要素数1の場合に限りプリミティブを返す
 *
 * 返却形式:
 * - 単一行: TokenArray（例: ['a'] / [1,2,3]）
 * - 複数行: TokenMatrix（例: [['a','b'],['c','d']]）
 * - primitive: true かつ単一行・要素数1のときのみ Token（例: 'a' / 1）
 *
 * @param {string} input 入力文字列
 * @param {object} [options] オプション
 * @param {boolean} [options.primitive=false] 単一行・単一要素のとき、配列ではなくプリミティブを返す
 * @returns {Token|TokenArray|TokenMatrix}
 *
 * @example
 * splitTokens('a');                         // ['a']
 * splitTokens('a', { primitive: true });    // 'a'
 * splitTokens('1 2 3');                     // [1,2,3]
 * splitTokens('a b\nc d');                  // [['a','b'],['c','d']]
 */
export default function splitTokens(input, { primitive = false } = {}) {
  const normalized = normalizeInput(input);
  const isSingleLine = !normalized.includes('\n');

  // 単一行か複数行で処理を分岐
  const result = isSingleLine ? parseSingleLine(normalized) : parseMultiLine(normalized);

  // オプションかつ単一値はプリミティブ値に変換
  return primitive && isSingleLine && result.length === 1 ? result[0] : result;
}

File: packages/splitTokens-py/src/splitTokens.py
import re


# 入力文字列を正規化
def normalizeInput(s: str) -> str:
    s = re.sub(r"\r\n?", "\n", s)  # 改行統一
    s = re.sub(
        r"[\t\u3000,]", " ", s
    )  # 空白類（タブ・全角スペース・カンマ）を半角スペースに変換
    return s.strip()


# 文字型を数値型、小数型に変換
def convert_type(value):
    try:
        if "." in value:
            return float(value)
        else:
            return int(value)
    except ValueError:
        return value


# 一行の文字列を一次元配列に変換
def parseSingleLine(s: str):
    return [convert_type(token) for token in s.split()]


# 複数行の文字列を二次元配列に変換
def parseMultiLine(s: str):
    return [parseSingleLine(line) for line in s.splitlines() if line.strip()]


# 変換関数
def splitTokens(input):
    normalized = normalizeInput(input)
    isSingleLine = "\n" not in normalized
    # 単一行か複数行で処理を分岐
    return parseSingleLine(normalized) if isSingleLine else parseMultiLine(normalized)
